packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useEffect,
+	useState,
+	useRef,
+	useMemo,
+	useCallback,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+// const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+// 	useLayoutEffect : useEffect;
+const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+	useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+export default function useSelect( mapSelect, nextProps ) {
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const [ mapOutput, setMapOutput ] = useState( mapSelect( registry.select ) );
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+
+	const latestMapSelect = useRef( mapSelect );
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef( mapOutput );
+	const isMounted = useRef();
+
+	/**
+	 * Set current values.
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestMapSelect.current = mapSelect;
+		isMounted.current = true;
+	} );
+
+	/**
+	 * Flush queue if isAsync changes
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestIsAsync.current = isAsync;
+		renderQueue.flush( queueContext );
+	}, [ isAsync ] );
+
+	/**
+	 * Callback for doing changes to store.
+	 */
+	const onStoreChange = useCallback( () => {
+		if ( isMounted.current ) {
+			const newMapOutput = latestMapSelect.current( registry.select, registry );
+			if ( ! isShallowEqualObjects( latestMapOutput.current, newMapOutput ) ) {
+				latestMapOutput.current = newMapOutput;
+				setMapOutput( newMapOutput );
+			}
+		}
+	}, [ registry, nextProps ] );
+
+	/**
+	 * Map output should always be called if incoming prop dependencies
+	 * has changed.
+	 */
+	useIsomorphicLayoutEffect( () => {
+		onStoreChange();
+	}, [ registry, nextProps ] );
+
+	useIsomorphicLayoutEffect( () => {
+		const unsubscribe = registry.subscribe( () => {
```

Comment:
I expect this will break the re-rendering order, because the children will subcribe before the parents. Right?
I think the solution might be to `subscribe` synchronously at the render level (but only manage to do it once).

Reply:
hmm...

Reply:
Yeah, see https://react-redux.js.org/next/api/hooks#stale-props-and-zombie-children

Reply:
hmm so the subscription callback would likely have to be a ref maybe?

Reply:
This basically means for me that subscriptions should happen at the root level of the hook and not inside a `useEffect`.

Something like

```js
const registry = useRegistry();
const isAsync = useAsyncMode();
const previousRegistry = usePrevious( registry );
const previousIsAsync = usePrevious( isAsync );

if ( registry !== previousRegistry || isAsync !== previousIsAsync ) {
   // Trigger subscription
}
```

Reply:
That breaks React Suspense. They outline workarounds in the link I pasted.

Reply:
Curious to know how would this break Suspense?

Reply:
Side effects in renders that might be interrupted, blocked, batched, etc.

Reply:
yes, but this is very specific and if we can deal with it at the framework level, the better it is. I mean, we can ensure that the subscription callback is only run after the `useEffect` has triggered (using a ref or something).

Reply:
That's not specific. The subscription is the side effect.

Reply:
okay so the latest commit (69f3b80) makes onStoreChange dependent _only_ on `registry` and yet still uses the latest mapSelect via the ref.  This allows us to make it a dependency on the subscribe call here.  As far as I can tell, the changes mean we are now replicating what the original `withSelect` was done in terms of lifecycle (where change in registry means there's un unsubscribe on the old registry and a resubscribe on the new.

Reply:
To be honest, the solutions from the link you shared don't seem great either. How can we solve this in the hook itself?

Reply:
Hooks have no way of knowing where in the render tree they are called, so you can't without wrapping the component in an HOC.

Reply:
Are you aware of any React related discussions/docs about these matters?

Reply:
https://gist.github.com/markerikson/faac6ae4aca7b82a058e13216a7888ec

https://github.com/reduxjs/react-redux/issues/1179

Reply:
Do you think you'd be able to provide an example (codepen or something) where the React Async rendering would break this. Asking because that would be a good way to try and test alternative implementations addressing this issue.

Reply:
That would take a while to set up. The entire React API relies on render being idempotent. Note that the current implementation of `withSelect` also has this issue as it has side effects in the constructor.

From the React docs:

Conceptually, React does work in two phases:

- The render phase determines what changes need to be made to e.g. the DOM. During this phase, React calls render and then compares the result to the previous render.
- The commit phase is when React applies any changes. (In the case of React DOM, this is when React inserts, updates, and removes DOM nodes.) React also calls lifecycles like componentDidMount and componentDidUpdate during this phase.

The commit phase is usually very fast, but rendering can be slow. For this reason, the upcoming async mode (which is not enabled by default yet) breaks the rendering work into pieces, pausing and resuming the work to avoid blocking the browser. This means that React may invoke render phase lifecycles more than once before committing, or it may invoke them without committing at all (because of an error or a higher priority interruption).

Render phase lifecycles include the following class component methods:

- constructor
- componentWillMount
- componentWillReceiveProps
- componentWillUpdate
- getDerivedStateFromProps
- shouldComponentUpdate
- render
- setState updater functions (the first argument)

Because the above methods might be called more than once, it’s important that they do not contain side-effects. Ignoring this rule can lead to a variety of problems, including memory leaks and invalid application state. Unfortunately, it can be difficult to detect these problems as they can often be non-deterministic.

Strict mode can’t automatically detect side effects for you, but it can help you spot them by making them a little more deterministic. This is done by intentionally double-invoking the following methods:

- Class component constructor method
- The render method
- setState updater functions (the first argument)
- The static getDerivedStateFromProps lifecycle

>Note:
This only applies to development mode. Lifecycles will not be double-invoked in production mode.

For example, consider the following code:

```js
class TopLevelRoute extends React.Component {
  constructor(props) {
    super(props);

    SharedApplicationState.recordEvent('ExampleComponent');
  }
}
```

At first glance, this code might not seem problematic. But if `SharedApplicationState.recordEvent` is not idempotent, then instantiating this component multiple times could lead to invalid application state. This sort of subtle bug might not manifest during development, or it might do so inconsistently and so be overlooked.

By intentionally double-invoking methods like the component constructor, strict mode makes patterns like this easier to spot.

Reply:
Thanks for sharing. I think we probably have a lot of components that suffer from these issues. Makes me think we don't really support async mode. Not to say that we shouldn't but if feels like something we should invest in at a more global level.

From what I understand, say we have a global counter, we could increment (side effect) from within hooks. If the render function increments this counter, we'll still be certain that children components have pick a higher-number from this counter than their parent as even if react could double invoke parents render function ... it will still call the render of the children after the parents or is this a wrong assumption (assuming initial rendering)?

Reply:
Here is the execution order:

```
render parent
render childA
render childB
useLayoutEffect cleanup childA
useLayoutEffect cleanup childB
useLayoutEffect cleanup parent
useLayoutEffect effect childA
useLayoutEffect effect childB
useLayoutEffect effect parent
useEffect cleanup childA
useEffect effect childA
useEffect cleanup childB
useEffect effect childB
useEffect cleanup parent
useEffect effect parent
```

You can't run side effects in a parent before running them in its children, because side effects only happen after the first render. This is true without hooks as well. The only alternative I see would be to stagger-render (pause render at each level of the tree to run side effects), but performance would suffer, and you would still have issues when children are rendered conditionally/dynamically.